drivers/SmartThings/philips-hue/src/init.lua

```diff
@@ -1554,6 +1564,39 @@ if hue.datastore["dni_to_device_id"] == nil then
   hue.datastore["dni_to_device_id"] = {}
 end
 
+
+if hue.datastore["api_keys"] == nil then
+  hue.datastore["api_keys"] = {}
+end
+
+if Discovery.api_keys and next(Discovery.api_keys) ~= nil then
+  local datastore_dirty = false
+  for k, v in pairs(Discovery.api_keys) do
+    hue.datastore.api_keys[k] = v
+    datastore_dirty = true
+  end
+
+  if datastore_dirty then
+    hue.datastore:save()
+  end
+end
+
+Discovery.api_keys = setmetatable({}, {
```

Comment:
Metatables used like this is what makes people hate lua...It is counterintuitive to reach into a different component's fields, and modify its functionality from outside the component. I see why it was done here, and you're comment in disco.lua helps sort things out, but its pretty hacky. Ill approve with this as is though, since i dont think there is a way to do it  without many changes to disco.lua.

Reply:
Oh I hate this with a passion. But it's a minimum-entropy fix so that I can get back to finishing the integration tests so that I can have regression coverage so that I can refactor the driver.